### PR TITLE
Skip joined entity creation for empty relation (#10889)

### PR DESCRIPTION
**Pull request by noemi-salaun**
(The following is translated from french with Google Translate, I'm sorry :/ )

In case the root entity has its primary key as a relationship and the joined entity has a simple primary key, the RSM orders the ID of the joined entity before having the ID which defined the root entity.

The ObjectHydrator therefore first tries to hydrate the joined entity. We then fall into a special case where the parent (the root entity) has not yet been created and the hydrator tries to start by creating the child, except that in our case, the child is supposed to be null.

I therefore added the `isset($nonemptyComponents[$dqlAlias])` check which we find further in the code and which is made to skip the creation of the joined entity when the parent is already hydrated. I think it makes sense to do the same check in the case where the hydrator starts with the joined entity.

### CODE REVIEW

**Comment by mpdude** [comment]
What does it mean when the `GH10889Hand` class has an `@Id` column that is a one to one association and a generated value at the same time?

**Comment by noemi-salaun** [comment]
ah you're right it's an oversight. During my tests I wanted to see how the ObjectHydrator behaved with a simple ID rather than a relationship, and I left an annotation. It doesn't change anything about the problem here but I'm going to remove it ASAP

**Comment by noemi-salaun** [comment]
@mpdude It's corrected. I checked, I still have the same problem without the `GeneratedValue` annotation, and the problem is still fixed with my changes.

**Comment by noemi-salaun** [comment]
I don't think the failed psalm static analysis is related to my changes

**Comment by noemi-salaun** [comment]
@mpdude 
Any info on when these changes might be merge? I am still locked to v2.15 in my project because of this

**Comment by Prokyonn** [comment]
We've tested the changes from this PR in Sulu, everything seems to be running without encountering any issues. 

However, until this is merged, we also have to lock the doctrine version in our customer projects to v2.15. Given this situation, we would greatly appreciate it if you could prioritise merging this PR.

**Comment by derrabus** [comment]
> if you could prioritise merging this PR.

The PR cannot be merged due to conflicts.

**Comment by winiarekk** [comment]
@noemi-salaun will you be able to resolve these conflicts?

**Comment by noemi-salaun** [comment]
@derrabus I just rebased my work on the 2.19.x branch, it should be good now

**Comment by noemi-salaun** [comment]
@mpdude @derrabus
Is there still something blocking this PR?

File: tests/Tests/ORM/Functional/Ticket/GH10889Test.php
Comment (greg0ire) [inline tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: Ok so a hand is linked to one user… makes sense. And… one user… has only one hand? So we are dealing with a one-handed species… weird but OK.

File: tests/Tests/ORM/Functional/Ticket/GH10889Test.php
Comment (greg0ire) [inline tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: Many hands share the same thumb?

File: tests/Tests/ORM/Functional/Ticket/GH10889Test.php
Comment (noemi-salaun) [reply tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: My tests started with more finger, but I simplified it more and more to try to isolate the problem as much as possible

Comment (greg0ire) [reply tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: Ok… well maybe you will able to find an example that makes a little more sense? Especially the thumb part is confusing.

Comment (noemi-salaun) [reply tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: Are you not focusing on the wrong thing here? 😅

**Comment by noemi-salaun** [comment]
@greg0ire I changed the test with entities with less meaning

File: tests/Tests/ORM/Functional/Ticket/GH10889Test.php
Comment (greg0ire) [inline tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: Many boxes have one item?

Comment (noemi-salaun) [reply tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: Yes, the same item can be put in the box of different user.
Why are you so picky for this example? It's just here to mimic a real case while only focusing on the issue. I can clutter it with ten more properties to make it an interesting piece of software but that does not serve the point in my opinion.
Or should I call them class A B and C so we can move on with the real issue I'm trying so solve here? ^^

Comment (greg0ire) [reply tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: My hope is that if I can make sense if the test data, I'll be able to make sense of the test, and then the source code change. And if you can't be think of 2 relationships that, you know, make sense, then use A, B, C, at least it is not misleading. Anyway, a user could have a single contract with an internet provider.

Comment (noemi-salaun) [reply tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: Ok I see. But my new example works fine. Each user have 1 box where they can store 1 item of their choice. Multiple users can store the same item (like item you put in a cart)

Comment (greg0ire) [reply tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: I'll try to keep that in mind when giving this another review, but I would rather not.

Comment (greg0ire) [reply tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: Example that I think can work:
- Person
- Resume
- Company

A Person has 1 Resume, and that Resume could have a `currentCompany` field, nullable, and allowing only one company.

Comment (noemi-salaun) [reply tests/Tests/ORM/Functional/Ticket/GH10889Test.php]: @greg0ire I changed it with your example :)

**Review by SenseException — APPROVED** [review]

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @noemi-salaun !

**Pull request merged**
